man/common/authorizations.md

```diff
@@ -9,21 +9,38 @@ Authorization for use of an object in TPM2.0 can come in 3 different forms:
 
 ## Passwords
 
-Passwords are interpreted in three forms; string, hex-string or a file. A string password is not
-interpreted, and is directly used for authorization. A hex-string password is converted from
-a hexidecimal form into a byte array form, thus allowing passwords with non-printable
-and/or terminal un-friendly characters.
-A file form should be the path of a file containing a password in string or hex-string format to be read by the tool.
-Storing passwords in files prevents information leakage, passwords passed as options can be read from the process list.
-
-By default passwords are assumed to be in the string form. Password form is specified
-with special prefix values, they are:
-
-  * str: - Used to indicate it is a raw string. Useful for escaping a password that starts
-         with the "hex:" prefix.
-  * hex: - Used when specifying a password in hex string format.
-  * file: - Used when specifying a password stored in a file. Useful to prevent leaking the
-         password to UNIX utilities (such as ps).
+Passwords are interpreted in four forms:
+1. String
+2. Hex-string
+3. File
+4. Stdin
+
+A string password is not interpreted, and is directly used for authorization. A
+hex-string password is converted from a hexidecimal form into a byte array form,
+thus allowing passwords with non-printable and/or terminal un-friendly
+characters. A file form should be the path of a file containing a password in
+string or hex-string format to be read by the tool. Storing passwords in files
+prevents information leakage, passwords passed as options can be read from the
+process list. Stdin password input, just like string, is not interpreted. When
+specifying stdin passwords, there are two ways: Prompt and Process substitution.
+When using the prompt method, specifying the password should be followed with
+a return and ctrl+d so it becomes equivalent to a string password. As an example,
+tpm2_tool -p str:password is equivalent to tpm2_tool -p -, followed by
+prompt:password<return><ctrl+d>
+
+By default passwords are assumed to be in the string form. Password form is
+specified with special prefix values, they are:
+
+  * str:<password> Used to indicate it is a raw string. Useful for escaping a
+                   password that starts with the "hex:" prefix.
+  * hex:<password> Used when specifying a password in hex string format.
+  * file:<file>    Used when specifying a password stored in a file. Useful to
+                   prevent leaking the password to UNIX utilities (such as ps).
+  * "-"            Used when specifying a password from stdin. There are 2 ways
+                   to specify the password. A prompt and by the process
+                   substitution method. Note that this is is similar to using
+                   file:-, however this option additionally hides password
+                   characters being echoed back to screen when using prompt.
```

Comment:
Huh, I don't understand, if we just fread(stdin, ...), who would be echoing them back, I don't understand why we need termios?

Reply:
Can't you just use getpass from gnulibc?

```
man getpass
NAME
       getpass - get a password

SYNOPSIS
       #include <unistd.h>

       char *getpass(const char *prompt);
```

Reply:
> Huh, I don't understand, if we just fread(stdin, ...), who would be echoing them back, I don't understand why we need termios?

This is the case where the user is keying in the pswd and auth_util reads it as stdin. In this case user key presses are echoed back.

Reply:
> Can't you just use getpass from gnulibc?
> 
> ```
> man getpass
> NAME
>        getpass - get a password
> 
> SYNOPSIS
>        #include <unistd.h>
> 
>        char *getpass(const char *prompt);
> ```

Yes we could, however we will have to add something other than '-' to the auth language. Right now we are telling apart stdin from prompt using termios instead of getpass

Reply:
File:- use studio, just - use getpass

Reply:
Oh I am an idiot, getpass is deprecated and they want us to go to termios per the manpage...